README.rst
=======================================
Django Microsoft Authentication Backend
=======================================


.. image:: https://img.shields.io/pypi/v/django_microsoft_auth.svg
    :target: https://pypi.python.org/pypi/django_microsoft_auth
    :alt: PyPi

.. image:: https://img.shields.io/pypi/pyversions/django_microsoft_auth.svg
    :target: https://pypi.python.org/pypi/django_microsoft_auth
    :alt: Python Versions

.. image:: https://travis-ci.org/AngellusMortis/django_microsoft_auth.svg?branch=master
    :target: https://travis-ci.org/AngellusMortis/django_microsoft_auth/
    :alt: Travis CI

.. image:: https://readthedocs.org/projects/django-microsoft-auth/badge/?version=latest
    :target: https://django-microsoft-auth.readthedocs.io/en/latest/?badge=latest
    :alt: Documentation

.. image:: https://pyup.io/repos/github/AngellusMortis/django_microsoft_auth/shield.svg
    :target: https://pyup.io/repos/github/AngellusMortis/django_microsoft_auth/
    :alt: Updates

.. image:: https://coveralls.io/repos/github/AngellusMortis/django_microsoft_auth/badge.svg?branch=master
    :target: https://coveralls.io/github/AngellusMortis/django_microsoft_auth?branch=master
    :alt: Coverage

.. image:: https://api.codeclimate.com/v1/badges/ea41b61fa3a1e22e92e9/maintainability
   :target: https://codeclimate.com/github/AngellusMortis/django_microsoft_auth/maintainability
   :alt: Maintainability

.. image:: https://api.codeclimate.com/v1/badges/ea41b61fa3a1e22e92e9/test_coverage
   :target: https://codeclimate.com/github/AngellusMortis/django_microsoft_auth/test_coverage
   :alt: Test Coverage


Simple app to enable Microsoft Account, Office 365 and Xbox Live authentication
as a Django authentication backend.


* Free software: MIT license
* Documentation: https://django-microsoft-auth.readthedocs.io.

Features
--------

* Provides Django authentication backend to do Microsoft authentication
  (including Microsoft accounts, Office 365 accounts and Azure AD accounts)
  and Xbox Live authentication.

* Provides Microsoft OAuth client to interfacing with Microsoft accounts

Python/Django support
---------------------

`django_microsoft_auth` follows the same `support cycle as Django <https://www.djangoproject.com/download/#supported-versions>`_,
with one exception: no Python 2 support. If you absoutely need Python 2.7
support, everything should largely already work, but you may need to patch
`microsoft_auth.admin` and/or other files to get it to work.

Supported python versions:  3.5+

Supported Django version: 2.1+

https://docs.djangoproject.com/en/stable/faq/install/#what-python-version-can-i-use-with-django


Credits
-------

This package was created with Cookiecutter_ and the
`audreyr/cookiecutter-pypackage`_ project template.

.. _Cookiecutter: https://github.com/audreyr/cookiecutter
.. _`audreyr/cookiecutter-pypackage`: https://github.com/audreyr/cookiecutter-pypackage


HISTORY.rst
=======
History
=======

2.4.0 (2020-8-14)
-----------------

* Add support for Django 3.0 & 3.1
* Dropped support for Django 1.11
* Fixes error in name parsing
  - Fixes crash arising from missing comma in fullname
  - Assigns fullname to firstname if it can't be split
* Fixes installation with latest pip
* Fixes OpenID library crash due to uninitialized jwk

2.3.1 (2019-7-20)
-----------------

* Fixes W002 being displayed if you set the SITE_ID setting
* Fixes migration except for non-SQLite DBs

2.3.0 (2019-6-2)
-----------------

* Adds Django cache support for OIDC config/JWKS

2.2.3 (2019-5-19)
-----------------

* Django settings trump Constance settings always now
* Fixes `microsoft.conf.config` not using Constance in some cases

2.2.2 (2019-5-8)
----------------

* Fixes Javascript typo

2.2.1 (2019-4-28)
-----------------

* Namespaces messages passed back as part of login

2.2.0 (2019-3-26)
-----------------

* Adds new setting for callback hook right before `auth_callback` view renders
  to override context data
* Adds example non-admin login form example in test site
* Renames `admin_login.js` and `admin_login.css` to just `login.js` and
  `login.css`

2.1.1 (2019-3-24)
-----------------

* Adds profile back as a default scope since Microsoft is added if it is not

2.1.0 (2019-3-23)
-----------------

* Adds support for multiple SITE_IDs. If the setting is not provided, it will
  pull it from the request object
* Adds new setting for callback hook after
  `microsoft_auth.backends.MicrosoftAuthenticationBackend` authenticates user.

2.0.1 (2019-3-19)
-----------------

* Removes `profile` from required scopes

2.0.0 (2019-3-19)
-----------------

* Replaces deprecated Microsoft auth scopes with proper OpenID Connect ones
    * WARNING: Breaking change. New scopes provide a new user id. See
      `migration docs <https://django-microsoft-auth.readthedocs.io/en/latest/usage.html#migrating-from-1-0-to-2-0>`_
      for details.
* Pulls authorization/token URLs directly from Microsoft
* Adds id token validation
* Admin pages for the auth type that is not enable is disabled by default now.
  They can be re-enabled with `MICROSOFT_AUTH_REGISTER_INACTIVE_ADMIN = True`
* Extra scopes can be provieded via the `MICROSOFT_AUTH_EXTRA_SCOPES` setting
  (space delimited). These scopes are *added* to the default required scopes
  (`openid email` for Microsoft Auth and
  `XboxLive.signin XboxLive.offline_access` for Xbox Live auth)

1.3.3 (2019-3-16)
-----------------

* Adds expiration to state values (hardcoded 5 minutes)

1.3.2 (2019-3-16)
-----------------

* Changes state validation to use cryptographic signing now. State
  validation should be signfincally more relaible now.

1.3.1 (2019-3-16)
-----------------

* Adds more logging around CSRF/State failures

1.3.0 (2019-3-5)
----------------

* Adds support for other tenant IDs for Microsoft
  authentication (thanks aviv)

1.2.1 (2019-2-28)
-----------------

* Adds missing migration for changing `microsoft_id` from 32 to 36 length

1.2.0 (2019-1-13)
-----------------

* Adds various checks and logging to validate setup to help with debugging
* Adds support for `http://localhost` as a redirect URI base if `DEBUG` is
  enabled
* Fixes Javascript message passing if using a non-standard port (something
  other than 80 or 443)

1.1.0 (2018-7-3)
----------------
* Removes o365 option. New authorization URL works well enough for both
    * Xbox Live Auth still uses old Microsoft Auth URL
    * 'o365' will still work as a MICROSOFT_AUTH_LOGIN_TYPE value,
      but you should change it to 'ma'
* Adds extras:
    * `ql`: DjangoQL package and support
    * `test`: test dependencies (same as test_requires packages)
    * `dev`: `ql`+`test` and extra dev only dependencies like
      `twine` and `pip-tools`
* Pip 10 support (thanks Shigumitsu!)
* Fixes max length of o365 IDs (thanks Shigumitsu!)

1.0.6 (2018-4-8)
----------------
* Added patched username validator to allow spaces for usernames for
  Xbox Live Gamertags

1.0.5 (2018-4-8)
----------------
* Added missing templates and static files to MANIFEST

1.0.4 (2017-12-2)
-----------------

* Updated Django category to include 2.0

1.0.3 (2017-12-2)
-----------------

* Updated for Django 2.0

1.0.2 (2017-11-27)
------------------

* Changed Development Status category to Stable

1.0.0 (2017-11-19)
------------------

* First release on PyPI.


microsoft_auth/client.py
import json
import logging

import jwt
import requests
from django.contrib.sites.models import Site
from django.core.cache import cache
from django.urls import reverse
from jwt.algorithms import RSAAlgorithm
from requests_oauthlib import OAuth2Session

from .conf import (
    CACHE_KEY_JWKS,
    CACHE_KEY_OPENID,
    CACHE_TIMEOUT,
    LOGIN_TYPE_XBL,
)
from .utils import get_scheme

logger = logging.getLogger("django")


class MicrosoftClient(OAuth2Session):
    """ Simple Microsoft OAuth2 Client to authenticate them

        Extended from Requests-OAuthlib's OAuth2Session class which
            does most of the heavy lifting

        https://requests-oauthlib.readthedocs.io/en/latest/

        Microsoft OAuth documentation can be found at
        https://developer.microsoft.com/en-us/graph/docs/get-started/rest
    """

    _config_url = "https://login.microsoftonline.com/{tenant}/v2.0/.well-known/openid-configuration"  # noqa

    _xbox_authorization_url = "https://login.live.com/oauth20_authorize.srf"
    _xbox_token_url = (
        "https://user.auth.xboxlive.com/user/authenticate"  # nosec
    )
    _profile_url = "https://xsts.auth.xboxlive.com/xsts/authorize"

    xbox_token = {}

    config = None

    # required OAuth scopes
    SCOPE_XBL = ["XboxLive.signin", "XboxLive.offline_access"]
    SCOPE_MICROSOFT = ["openid", "email", "profile"]

    def __init__(self, state=None, request=None, *args, **kwargs):
        from .conf import config

        self.config = config

        extra_scopes = self.config.MICROSOFT_AUTH_EXTRA_SCOPES

        try:
            current_site = Site.objects.get_current(request)
        except Site.DoesNotExist:
            current_site = Site.objects.first()

        domain = current_site.domain
        path = reverse("microsoft_auth:auth-callback")
        scope = " ".join(self.SCOPE_MICROSOFT)

        if self.config.MICROSOFT_AUTH_LOGIN_TYPE == LOGIN_TYPE_XBL:
            scope = " ".join(self.SCOPE_XBL)

        scope = "{} {}".format(scope, extra_scopes).strip()

        scheme = get_scheme(request, self.config)

        super().__init__(
            self.config.MICROSOFT_AUTH_CLIENT_ID,
            scope=scope,
            state=state,
            redirect_uri="{0}://{1}{2}".format(scheme, domain, path),
            *args,
            **kwargs
        )

    @property
    def openid_config(self):
        config = cache.get(CACHE_KEY_OPENID)

        if config is None:
            config_url = self._config_url.format(
                tenant=self.config.MICROSOFT_AUTH_TENANT_ID
            )
            response = self.get(config_url)

            if response.ok:
                config = response.json()
                cache.set(CACHE_KEY_OPENID, config, CACHE_TIMEOUT)

        return config

    @property
    def jwks(self):
        jwks = cache.get(CACHE_KEY_JWKS, [])

        if len(jwks) == 0:
            jwks_uri = self.openid_config["jwks_uri"]
            if jwks_uri is None:
                return []

            response = self.get(jwks_uri)

            if response.ok:
                jwks = response.json()["keys"]
                cache.set(CACHE_KEY_JWKS, jwks, CACHE_TIMEOUT)
        return jwks

    def get_claims(self, allow_refresh=True):
        if self.token is None:
            return None

        token = self.token["id_token"].encode("utf8")

        kid = jwt.get_unverified_header(token)["kid"]
        jwk = None
        public_key = None
        for key in self.jwks:
            if kid == key["kid"]:
                jwk = key
                break

        if jwk is None:
            if allow_refresh:
                logger.warn(
                    "could not find public key for id_token, "
                    "refreshing OIDC config"
                )
                cache.delete(CACHE_KEY_JWKS)
                cache.delete(CACHE_KEY_OPENID)

                return self.get_claims(allow_refresh=False)
            else:
                logger.warn("could not find public key for id_token")
                return None

        public_key = RSAAlgorithm.from_jwk(json.dumps(jwk))

        try:
            claims = jwt.decode(
                token,
                public_key,
                algoithm="RS256",
                audience=self.config.MICROSOFT_AUTH_CLIENT_ID,
            )
        except jwt.PyJWTError as e:
            logger.warn("could verify id_token sig: {}".format(e))
            return None

        return claims

    def authorization_url(self):
        """ Generates Microsoft/Xbox or a Office 365 Authorization URL """

        auth_url = self.openid_config["authorization_endpoint"]
        if self.config.MICROSOFT_AUTH_LOGIN_TYPE == LOGIN_TYPE_XBL:
            auth_url = self._xbox_authorization_url

        return super().authorization_url(auth_url, response_mode="form_post")

    def fetch_token(self, **kwargs):
        """ Fetchs OAuth2 Token with given kwargs"""

        return super().fetch_token(  # pragma: no cover
            self.openid_config["token_endpoint"],
            client_secret=self.config.MICROSOFT_AUTH_CLIENT_SECRET,
            **kwargs
        )

    def fetch_xbox_token(self):
        """ Fetches Xbox Live Auth token.

            token must contain a valid access_token
                - retrieved from fetch_token

            Reversed engineered from existing Github repos,
                no "official" API docs from Microsoft

            Response will be similar to
            {
                'Token': 'token',
                'IssueInstant': '2016-09-27T15:01:45.225637Z',
                'DisplayClaims': {'xui': [{'uhs': '###################'}]},
                'NotAfter': '2016-10-11T15:01:45.225637Z'
            }
        """

        # Content-type MUST be json for Xbox Live
        headers = {
            "Content-type": "application/json",
            "Accept": "application/json",
        }
        params = {
            "RelyingParty": "http://auth.xboxlive.com",
            "TokenType": "JWT",
            "Properties": {
                "AuthMethod": "RPS",
                "SiteName": "user.auth.xboxlive.com",
                "RpsTicket": "d={}".format(self.token["access_token"]),
            },
        }
        response = requests.post(
            self._xbox_token_url, data=json.dumps(params), headers=headers
        )

        if response.status_code == 200:
            self.xbox_token = response.json()

        return self.xbox_token

    def get_xbox_profile(self):
        """
            Fetches the Xbox Live user profile from Xbox servers

            xbox_token must contain a valid Xbox Live token
                - retrieved from fetch_xbox_token

            Reversed engineered from existing Github repos,
                no "official" API docs from Microsoft

            Response will be similar to
            {
                'NotAfter': '2016-09-28T07:19:21.9608601Z',
                'DisplayClaims': {
                    'xui': [
                        {
                            'agg': 'Adult',
                            'uhs': '###################',
                            'usr': '###',
                            'xid': '################',
                            'prv': '### ### ###...',
                            'gtg': 'Gamertag'}]},
                'IssueInstant': '2016-09-27T15:19:21.9608601Z',
                'Token': 'token'}
        """

        if "Token" in self.xbox_token:
            # Content-type MUST be json for Xbox Live
            headers = {
                "Content-type": "application/json",
                "Accept": "application/json",
            }
            params = {
                "RelyingParty": "http://xboxlive.com",
                "TokenType": "JWT",
                "Properties": {
                    "UserTokens": [self.xbox_token["Token"]],
                    "SandboxId": "RETAIL",
                },
            }
            response = requests.post(
                self._profile_url, data=json.dumps(params), headers=headers
            )

            if response.status_code == 200:
                return response.json()["DisplayClaims"]["xui"][0]
        return {}

    def valid_scopes(self, scopes):
        """ Validates response scopes based on MICROSOFT_AUTH_LOGIN_TYPE """

        scopes = set(scopes)
        required_scopes = None
        if self.config.MICROSOFT_AUTH_LOGIN_TYPE == LOGIN_TYPE_XBL:
            required_scopes = set(self.SCOPE_XBL)
        else:
            required_scopes = set(self.SCOPE_MICROSOFT)

        # verify all require_scopes are in scopes
        return required_scopes <= scopes
